### ISSUE TITLE
IEDriverServer.exe crashes when trying to add a cookie

### ISSUE THREAD

**Issue by icnocop** [SeleniumHQ/selenium#7122]
## 🐛 Bug Report

When I try to add a cookie to the cookie container, IEDriverServer.exe crashes.

## To Reproduce

See the test below.

## Expected behavior

I expected the cookie to be added to the cookie container.

## Test script or set of commands reproducing this issue

```
using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using OpenQA.Selenium;
using OpenQA.Selenium.IE;

[TestClass]
public class UnitTest1
{
    [TestMethod]
    public void TestMethod1()
    {
        var explorerOptions = new InternetExplorerOptions
        {
            EnsureCleanSession = true,
            IntroduceInstabilityByIgnoringProtectedModeSettings = false
        };

        var internetExplorerDriverService = InternetExplorerDriverService.CreateDefaultService();
        internetExplorerDriverService.HideCommandPromptWindow = false;
        internetExplorerDriverService.LoggingLevel = InternetExplorerDriverLogLevel.Trace;
        internetExplorerDriverService.LogFile = @"C:\selenium.log";
        InternetExplorerDriver driver = new InternetExplorerDriver(internetExplorerDriverService, explorerOptions);

        driver.Url = "http://www.seleniumhq.org";
        driver.Navigate();
        var cookieContainer = driver.Manage().Cookies;

        Cookie cookie = new Cookie(
            "test",
            "0",
            null,
            "/",
            DateTime.MaxValue);

        var currentCookie = cookieContainer.GetCookieNamed(cookie.Name);
        if (currentCookie != null)
        {
            cookieContainer.DeleteCookieNamed(cookie.Name);
        }
        cookieContainer.AddCookie(cookie);
    }
}
```

Exception
```
OpenQA.Selenium.WebDriverException: A exception with a null response was thrown sending an HTTP request to the remote WebDriver server for URL http://localhost:4726/session/87e938dd-0509-48b4-a884-18a50685ade8/cookie. The status of the exception was ReceiveFailure, and the message was: The underlying connection was closed: An unexpected error occurred on a receive. ---> System.Net.WebException: The underlying connection was closed: An unexpected error occurred on a receive. ---> System.IO.IOException: Unable to read data from the transport connection: An existing connection was forcibly closed by the remote host. ---> System.Net.Sockets.SocketException: An existing connection was forcibly closed by the remote host
```

StackTrace
```
at System.Net.Sockets.Socket.Receive(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags)
   at System.Net.Sockets.NetworkStream.Read(Byte[] buffer, Int32 offset, Int32 size)
--- End of inner exception stack trace ---
    at System.Net.Sockets.NetworkStream.Read(Byte[] buffer, Int32 offset, Int32 size)
   at System.Net.PooledStream.Read(Byte[] buffer, Int32 offset, Int32 size)
   at System.Net.Connection.SyncRead(HttpWebRequest request, Boolean userRetrievedStream, Boolean probeRead)
--- End of inner exception stack trace ---
    at System.Net.HttpWebRequest.GetResponse()
   at OpenQA.Selenium.Remote.HttpCommandExecutor.MakeHttpRequest(HttpRequestInfo requestInfo)
--- End of inner exception stack trace ---
    at OpenQA.Selenium.Remote.HttpCommandExecutor.MakeHttpRequest(HttpRequestInfo requestInfo)
   at OpenQA.Selenium.Remote.HttpCommandExecutor.Execute(Command commandToExecute)
   at OpenQA.Selenium.Remote.DriverServiceCommandExecutor.Execute(Command commandToExecute)
   at OpenQA.Selenium.Remote.RemoteWebDriver.Execute(String driverCommandToExecute, Dictionary`2 parameters)
   at OpenQA.Selenium.Remote.RemoteCookieJar.AddCookie(Cookie cookie)
   at TestMethod1() in UnitTest1.cs:line 42
```

Windows Application Event Log Entry
```
Faulting application name: IEDriverServer.exe, version: 3.141.0.0, time stamp: 0x5bd9ea9a
Faulting module name: IEDCAE7.tmp, version: 3.141.0.0, time stamp: 0x5bd9ea91
Exception code: 0xc0000409
Fault offset: 0x000d89c5
Faulting process id: 0x1354
Faulting application start time: 0x01d4f725833dde76
Faulting application path: C:\Users\rami.abughazaleh\Documents\Visual Studio 2017\Projects\UnitTestProject5\bin\Debug\IEDriverServer.exe
Faulting module path: C:\Users\RAMI~1.ABU\AppData\Local\Temp\IEDCAE7.tmp
Report Id: 46754b0e-5605-477c-b949-c807dc6a23c5
Faulting package full name: 
Faulting package-relative application ID: 
```

[selenium.log](https://github.com/SeleniumHQ/selenium/files/3099868/selenium.log)

## Environment

OS: Windows 10
Browser: IE
Browser version: 11.437.17763.0
Browser Driver version: 3.141.0
Language Bindings version: C# .NET 4.7.1

It works in Chrome without issues.

Thank you.

**Edit**
I changed `IntroduceInstabilityByIgnoringProtectedModeSettings` to `false` and uploaded selenium.log.

**Comment by jimevans** [comment]
@icnocop Two things. First, you’re setting `IntroduceInstabilityByIgnoringProtectedModeSettings`. Please set that to false. Second, you’re going to all the trouble to create a trace-level log for the driver, please supply it.

**Comment by jimevans** [comment]
The issue here is the use of `DateTime.MaxValue`. At present, the IE driver cannot handle cookie expiration dates greater than December 31, 3000 23:59:59. There is a definite issue here, but the workaround is to use a reasonable cookie expiration date rather than December 31, 9999.

**Comment by icnocop** [comment]
I confirmed that using `new DateTime(3000, 12, 31, 23, 59, 59, 999, DateTimeKind.Utc)` instead of `DateTime.MaxValue` works!

Thank you for a work-around! 😄

**Comment by diemol** [comment]
Closing as the OP reports that it is working now.

**Comment by icnocop** [comment]
It's working only with a work-around.

@jimevans indicated that it's still an issue.

I'm thinking IEDriver should be updated so that at least it doesn't crash; it should probably validate the date and indicate what the maximum allowed date is if the input exceeds that value.

Should this issue be reopened?

**Comment by jimevans** [comment]
I have a local change to the IE driver that will do exactly that. In particular, it limits the max cookie age to 2^32 - 1 seconds beyond the time the cookie is being set (using the max-age attribute of the cookie). Do note, though, that the new max date for a cookie will be about 60 years from the current date, rather than the 31 December 3000 date.

**Comment by icnocop** [comment]
Great!

I can't seem to re-open this issue, but if it will be re-closed soon anyway, it's fine with me.

Thank you!

**Comment by jimevans** [comment]
@icnocop Fixed in 103f12a. Now the driver returns an error if the requested expiration date is larger than the max allowed, and the exception message tells the user what the maximum allowed cookie expiration date is. Note that maximum cookie expiration is now relative to the date/time when the cookie is being set, rather than an absolute, hard-coded future date. Please adjust your calling code accordingly.
